File: src/rx_app.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#include <net/if.h>
#include <sys/ioctl.h>
#include <sys/socket.h>

#include <linux/can.h>
#include <linux/can/raw.h>


int main(int argc, char *argv[])
{
	int s;
	int i;
	int n_bytes;
	struct sockaddr_can address;
	struct ifreq ifr;
	struct can_frame frame;

	printf("CAN Sockets Demo - Receive\n");

	if ((s = socket(PF_CAN, SOCK_RAW, CAN_RAW)) < 0) {
		perror("Socket");
		return 1;
	}

	strcpy(ifr.ifr_name, "vcan0");
	ioctl(s, SIOCGIFINDEX, &ifr);

	memset(&address, 0, sizeof(address));
	address.can_family = AF_CAN;
	address.can_ifindex = ifr.ifr_ifindex;
	if (bind(s, (struct sockaddr *)&address, sizeof(address))) {
		perror("Bind");
		return 1;
	}

	while(1) {
		n_bytes = read(s, &frame, sizeof(struct can_frame));

		if (n_bytes < 0) {
			perror("Read");
			return 1;
		}

        canid_t can_id = frame.can_id;

		switch(can_id) {
			case 0x101:
				printf("Known Command 1\n");
				printf("0x%03X [%d] ", frame.can_id, frame.can_dlc);
				printf("\r\n");
				break;
			case 0x102:
				printf("Known Command 2\n");
				printf("0x%03X [%d] ", frame.can_id, frame.can_dlc);
				printf("\r\n");
				break;
			case 0x103:
				printf("Known Command 3\n");
				printf("0x%03X [%d] ", frame.can_id, frame.can_dlc);
				printf("\r\n");
				break;
			case 0x104:
				printf("Known Command 4\n");
				printf("0x%03X [%d] ", frame.can_id, frame.can_dlc);
				printf("\r\n");
				break;
			case 0x106:
				printf("Known Command 5\n");
				printf("0x%03X [%d] ", frame.can_id, frame.can_dlc);
				printf("\r\n");
				break;
			case 0x107:
				printf("Known Command 6\n");
				printf("0x%03X [%d] ", frame.can_id, frame.can_dlc);
				printf("\r\n");
				break;
			case 0x108:
				printf("Known Command 7\n");
				printf("0x%03X [%d] ", frame.can_id, frame.can_dlc);
				printf("\r\n");
				break;
			case 0x109:
				printf("Known Command 8\n");
				printf("0x%03X [%d] ", frame.can_id, frame.can_dlc);
				printf("\r\n");
				break;
			case 0x113:
				printf("Known Command 9\n");
				printf("0x%03X [%d] ", frame.can_id, frame.can_dlc);
				printf("\r\n");
				break;
			case 0x114:
				printf("Known Command 10\n");
				printf("0x%03X [%d] ", frame.can_id, frame.can_dlc);
				printf("\r\n");
				break;
			case 0x125:
				printf("Known Command 11\n");
				printf("0x%03X [%d] ", frame.can_id, frame.can_dlc);
				printf("\r\n");
				break;
			case 0x157:
				printf("Known Command 12\n");
				printf("0x%03X [%d] ", frame.can_id, frame.can_dlc);
				printf("\r\n");
				break;
			case 0x159:
				printf("Known Command 13\n");
				printf("0x%03X [%d] ", frame.can_id, frame.can_dlc);
				printf("\r\n");
				break;
			case 0x167:
				printf("Known Command 14\n");
				printf("0x%03X [%d] ", frame.can_id, frame.can_dlc);
				printf("\r\n");
				break;
			case 0x172:
				printf("Known Command 15\n");
				printf("0x%03X [%d] ", frame.can_id, frame.can_dlc);
				printf("\r\n");
				break;
			case 0x182:
				printf("Known Command 16\n");
				printf("0x%03X [%d] ", frame.can_id, frame.can_dlc);
				printf("\r\n");
				break;
			case 0x183:
				printf("Known Command 17\n");
				printf("0x%03X [%d] ", frame.can_id, frame.can_dlc);
				printf("\r\n");
				break;
			case 0x184:
				printf("Known Command 18\n");
				printf("0x%03X [%d] ", frame.can_id, frame.can_dlc);
				printf("\r\n");
				break;
			case 0x185:
				printf("Known Command 19\n");
				printf("0x%03X [%d] ", frame.can_id, frame.can_dlc);
				printf("\r\n");
				break;
			case 0x187:
				printf("Known Command 20\n");
				printf("0x%03X [%d] ", frame.can_id, frame.can_dlc);
				printf("\r\n");
				break;
			case 0x189:
				printf("Known Command 21\n");
				printf("0x%03X [%d] ", frame.can_id, frame.can_dlc);
				printf("\r\n");
				break;
			case 0x192:
				printf("Known Command 22\n");
				printf("0x%03X [%d] ", frame.can_id, frame.can_dlc);
				printf("\r\n");
				break;
			case 0x195:
				printf("Known Command 23\n");
				printf("0x%03X [%d] ", frame.can_id, frame.can_dlc);
				printf("\r\n");
				break;
			case 0x198:
				printf("Known Command 24\n");
				printf("0x%03X [%d] ", frame.can_id, frame.can_dlc);
				printf("\r\n");
				break;
			case 0x200:
				printf("Known Command 25\n");
				printf("0x%03X [%d] ", frame.can_id, frame.can_dlc);
				printf("\r\n");
				break;
			case 0x201:
				printf("Known Command 26\n");
				printf("0x%03X [%d] ", frame.can_id, frame.can_dlc);
				printf("\r\n");
				break;
			case 0x203:
				printf("Known Command 27\n");
				printf("0x%03X [%d] ", frame.can_id, frame.can_dlc);
				printf("\r\n");
				break;
			case 0x204:
				printf("Known Command 28\n");
				printf("0x%03X [%d] ", frame.can_id, frame.can_dlc);
				printf("\r\n");
				break;
			case 0x209:
				printf("Known Command 29\n");
				printf("0x%03X [%d] ", frame.can_id, frame.can_dlc);
				printf("\r\n");
				break;
			case 0x212:
				printf("Known Command 30\n");
				printf("0x%03X [%d] ", frame.can_id, frame.can_dlc);
				printf("\r\n");
				break;
			case 0x234:
				printf("Known Command 31\n");
				printf("0x%03X [%d] ", frame.can_id, frame.can_dlc);
				printf("\r\n");
				break;
			case 0x237:
				printf("Known Command 32\n");
				printf("0x%03X [%d] ", frame.can_id, frame.can_dlc);
				printf("\r\n");
				break;
			case 0x239:
				printf("Known Command 33\n");
				printf("0x%03X [%d] ", frame.can_id, frame.can_dlc);
				printf("\r\n");
				break;
			case 0x248:
				printf("Known Command 34\n");
				printf("0x%03X [%d] ", frame.can_id, frame.can_dlc);
				printf("\r\n");
				break;
			case 0x249:
				printf("Known Command 35\n");
				printf("0x%03X [%d] ", frame.can_id, frame.can_dlc);
				printf("\r\n");
				break;
			case 0x253:
				printf("Known Command 36\n");
				printf("0x%03X [%d] ", frame.can_id, frame.can_dlc);
				printf("\r\n");
				break;
			case 0x254:
				printf("Known Command 37\n");
				printf("0x%03X [%d] ", frame.can_id, frame.can_dlc);
				printf("\r\n");
				break;
			case 0x255:
				printf("Known Command 38\n");
				printf("0x%03X [%d] ", frame.can_id, frame.can_dlc);
				printf("\r\n");
				break;
			case 0x256:
				printf("Known Command 39\n");
				printf("0x%03X [%d] ", frame.can_id, frame.can_dlc);
				printf("\r\n");
				break;
			case 0x257:
				printf("Known Command 40\n");
				printf("0x%03X [%d] ", frame.can_id, frame.can_dlc);
				printf("\r\n");
				break;
			case 0x258:
				printf("Proxy: 1\n");
				printf("0x%03X [%d] ", frame.can_id, frame.can_dlc);
				printf("\r\n");
				break;
			case 0x452:
				printf("Proxy: 2\n");
				printf("0x%03X [%d] ", frame.can_id, frame.can_dlc);
				printf("\r\n");
				break;
			case 0x453:
				printf("Proxy: 3\n");
				printf("0x%03X [%d] ", frame.can_id, frame.can_dlc);
				printf("\r\n");
				break;
			case 0x554:
				printf("Non-memory operation: 1\n");
				printf("0x%03X [%d] ", frame.can_id, frame.can_dlc);
				printf("\r\n");
				break;
			case 0x555:
				printf("Non-memory operation: 5\n");
				printf("0x%03X [%d] ", frame.can_id, frame.can_dlc);
				printf("\r\n");
				break;
			case 0x556:
				printf("Non-memory operation: 10\n");
				printf("0x%03X [%d] ", frame.can_id, frame.can_dlc);
				printf("\r\n");
				break;
			case 0x557:
				printf("Non-memory operation: 100\n");
				printf("0x%03X [%d] ", frame.can_id, frame.can_dlc);
				printf("\r\n");
				break;
			case 0x558:
				printf("Non-memory operation: 1000\n");
				printf("0x%03X [%d] ", frame.can_id, frame.can_dlc);
				printf("\r\n");
				break;
			case 0x654:
				printf("Memory operation: 1\n");
				printf("0x%03X [%d] ", frame.can_id, frame.can_dlc);
				printf("\r\n");
				break;
			case 0x655:
				printf("Memory operation: 5\n");
				printf("0x%03X [%d] ", frame.can_id, frame.can_dlc);
				printf("\r\n");
				break;
			case 0x656:
				printf("Memory operation: 10\n");
				printf("0x%03X [%d] ", frame.can_id, frame.can_dlc);
				printf("\r\n");
				break;
			case 0x657:
				printf("Memory operation: 100\n");
				printf("0x%03X [%d] ", frame.can_id, frame.can_dlc);
				printf("\r\n");
				break;
			case 0x658:
				printf("Memory operation: 1000\n");
				printf("0x%03X [%d] ", frame.can_id, frame.can_dlc);
				printf("\r\n");
				break;
			case 0x754:
				printf("Flash operation: 1\n");
				printf("0x%03X [%d] ", frame.can_id, frame.can_dlc);
				printf("\r\n");
				break;
			case 0x755:
				printf("Flash operation: 5\n");
				printf("0x%03X [%d] ", frame.can_id, frame.can_dlc);
				printf("\r\n");
				break;
			case 0x756:
				printf("Flash operation: 10\n");
				printf("0x%03X [%d] ", frame.can_id, frame.can_dlc);
				printf("\r\n");
				break;
			case 0x757:
				printf("Flash operation: 100\n");
				printf("0x%03X [%d] ", frame.can_id, frame.can_dlc);
				printf("\r\n");
				break;
			case 0x758:
				printf("Flash operation: 1000\n");
				printf("0x%03X [%d] ", frame.can_id, frame.can_dlc);
				printf("\r\n");
				break;
			default:
				break;
		}
	}

	if (close(s) < 0) {
		perror("Close");
		return 1;
	}

	return 0;
}

File: src/rx_app.py
import socket
import struct
import array
import kuksa_client
import can

#TODO test: test the setup, do I receive messages the same as the C file? test the kuksa-client can send to their server with the arbitrary values? 
# TODO next steps: receive the can message, update so we can identify which can message we receive, decode message, send the message arbitrarily to the server, 
# make sure that what we do is, we receive message via tx_app.c, and then we receive correctly via python file, decode message via python-can? send to kuksa server 

def process_can_message(msg):
    if msg.arbitration_id == 0x453:
        vehicle_speed = int.from_bytes(msg.data, byteorder='big') # TODO: check byteorder, check the arbitration_id, these are arbitrary
        print(f"Vehicle speed is {vehicle_speed} km/h")
    elif msg.arbitration_id == 0x102:
        # Handle other CAN IDs and data decoding here
        pass
    else:
        print(f"Unknown CAN ID: {msg.arbitration_id}")

def main():
    s = socket.socket(socket.AF_CAN, socket.SOCK_RAW, socket.CAN_RAW)
    
    try:
        s.bind(('vcan0',))
    except OSError as e:
        print(f"Bind error: {e}")
        return 1
    
    print("CAN Sockets Demo - Receive")
    
    while True:
        try:
            frame = s.recv(16)  # Assuming the maximum CAN frame size is 16 bytes sizeof(struct can_frame) = 16
        except OSError as e:
            print(f"Read error: {e}")
            return 1
        
        can_id, can_dlc, data = struct.unpack("<IB3x8s", frame)
        
        print(f"0x{can_id:03X} [{can_dlc}] ", end="")

        
        # Decode the received frame using python-can
        msg = can.Message(
            arbitration_id=can_id,
            data=data,
            dlc=can_dlc,
            is_extended_id=False  # Assuming standard CAN frames
        )
        
        # Call the message processing function
        process_can_message(msg)
        


        data_bytes = array.array('B', data)

        client = kuksa_client.KuksaClientThread(config={})
        client.start() #does this need args or anything? is the client already started via agl and then im just connecting? TODO: confirm
        client.authorize() #TODO: confirm if this is needed?

        value_to_send = { #TODO: fill these in with the received can msg, test first does this even send to the server? 
            "path": "Vehicle.Speed",  # Path to the value
            "value": 54.0,            # The actual value you want to send
            "timestamp": "2023-06-13T09:17:09.103507+00:00"  # Timestamp (optional)
        }

                # Send the value to the Kuksa server
        client.setValue("Vehicle.Speed", value_to_send)

        # Stop the client's thread when done
        client.stop()
        
        if can_id == 0x101:
            print("Known Command 1")
        elif can_id == 0x102:
            print("Known Command 2")
        elif can_id == 0x103:
            print("Known Command 3")
        elif can_id == 0x104:
            print("Known Command 4")
        elif can_id == 0x106:
            print("Known Command 5")
        elif can_id == 0x107:
            print("Known Command 6")
        elif can_id == 0x108:
            print("Known Command 7")
        elif can_id == 0x109:
            print("Known Command 8")
        elif can_id == 0x113:
            print("Known Command 9")
        elif can_id == 0x114:
            print("Known Command 10")
        elif can_id == 0x125:
            print("Known Command 11")
        elif can_id == 0x157:
            print("Known Command 12")
        elif can_id == 0x159:
            print("Known Command 13")
        elif can_id == 0x167:
            print("Known Command 14")
        elif can_id == 0x172:
            print("Known Command 15")
        elif can_id == 0x182:
            print("Known Command 16")
        elif can_id == 0x183:
            print("Known Command 17")
        elif can_id == 0x184:
            print("Known Command 18")
        elif can_id == 0x185:
            print("Known Command 19")
        elif can_id == 0x187:
            print("Known Command 20")
        elif can_id == 0x189:
            print("Known Command 21")
        elif can_id == 0x192:
            print("Known Command 22")
        elif can_id == 0x195:
            print("Known Command 23")
        elif can_id == 0x198:
            print("Known Command 24")
        elif can_id == 0x200:
            print("Known Command 25")
        elif can_id == 0x201:
            print("Known Command 26")
        elif can_id == 0x203:
            print("Known Command 27")
        elif can_id == 0x204:
            print("Known Command 28")
        elif can_id == 0x209:
            print("Known Command 29")
        elif can_id == 0x212:
            print("Known Command 30")
        elif can_id == 0x234:
            print("Known Command 31")
        elif can_id == 0x237:
            print("Known Command 32")
        elif can_id == 0x239:
            print("Known Command 33")
        elif can_id == 0x248:
            print("Known Command 34")
        elif can_id == 0x249:
            print("Known Command 35")
        elif can_id == 0x253:
            print("Known Command 36")
        elif can_id == 0x254:
            print("Known Command 37")
        elif can_id == 0x255:
            print("Known Command 38")
        elif can_id == 0x256:
            print("Known Command 39")
        elif can_id == 0x257:
            print("Known Command 40")
        elif can_id == 0x258:
            print("Proxy: 1")
        elif can_id == 0x452:
            print("Proxy: 2")
        elif can_id == 0x453:
            print("Proxy: 3")
        elif can_id == 0x554:
            print("Non-memory operation: 1")
        elif can_id == 0x555:
            print("Non-memory operation: 5")
        elif can_id == 0x556:
            print("Non-memory operation: 10")
        elif can_id == 0x557:
            print("Non-memory operation: 100")
        elif can_id == 0x558:
            print("Non-memory operation: 1000")
        elif can_id == 0x654:
            print("Memory operation: 1")
        elif can_id == 0x655:
            print("Memory operation: 5")
        elif can_id == 0x656:
            print("Memory operation: 10")
        elif can_id == 0x657:
            print("Memory operation: 100")
        elif can_id == 0x658:
            print("Memory operation: 1000")
        elif can_id == 0x754:
            print("Flash operation: 1")
        elif can_id == 0x755:
            print("Flash operation: 5")
        elif can_id == 0x756:
            print("Flash operation: 10")
        elif can_id == 0x757:
            print("Flash operation: 100")
        elif can_id == 0x758:
            print("Flash operation: 1000")
        else:
            print("Unknown Command")

if __name__ == "__main__":
    exit_code = main()
    exit(exit_code)
